service/controllers/helpers/users-helpers.js

```diff
@@ -176,3 +189,87 @@ export function removeOAuthConnection(user, provider) {
 				items[0].get('provider'));
 		});
 }
+
+export function doChangePassword(user, oldPassword, newPassword, privileged = false) {
+	const salt = bcrypt.genSaltSync(10);
+	const passwordHash = bcrypt.hashSync(newPassword, salt);
+
+	if (!PasswordStrengthRegex.test(newPassword)) {
+		return Bluebird.reject(new WeakPasswordError('New password did not meet strength criteria'));
+	}
+
+	if (!privileged && !bcrypt.compareSync(oldPassword, user.passwordHash)) {
+		return Bluebird.reject(new BadPasswordError('Old password was incorrect.'));
+	}
+
+	return Users.updateAsync({ userId: user.userId, passwordHash: passwordHash });
+}
+
+export function doRequestPasswordReset(email) {
+	return getUserByEmail(email)
+		.then(user => {
+			if (user === null) {
+				return null;
+			}
+
+			if (!user.passwordHash) {
+				return null;
+			}
+
+			return Users.updateAsync({
+				userId: user.userId,
+				passwordResetToken: faker.random.alphaNumeric(20),
+				passwordResetExpiration: moment().add(1, 'd').toISOString()
+			});
+		})
+		.then(result => {
+			if (!result) {
+				return null;
+			}
+
+			return result.attrs;
+		});
+}
+
+export function doPerformPasswordReset(user, token, newPassword) {
+	if (!user.passwordResetToken || !user.passwordResetExpiration) {
+		return Bluebird.reject(new RejectedPasswordResetError(
+			`User "${user.userName}" has not requested a password reset.`));
+	}
+
+	if (user.passwordResetToken !== token) {
+		return Bluebird.reject(new RejectedPasswordResetError(
+			`Reset token did not match the one on file for user "${user.userName}".`));
+	}
+
+	if (moment().isAfter(user.passwordResetExpiration)) {
+		return Bluebird.reject(new RejectedPasswordResetError(
+			`Reset token has expired for user "${user.userName}".`));
+	}
+
+	const salt = bcrypt.genSaltSync(10);
+	const hash = bcrypt.hashSync(newPassword, salt);
+
+	return Users.updateAsync({
+		userId: user.userId,
+		passwordHash: hash,
+		passwordResetToken: null,
+		passwordResetExpiration: null })
+		.then(() => {
+			return true;
+		});
+}
+
+export function sanitizeUserInfo(user) {
+	const sanitized = _.pick(user, [
```

Comment:
It's not a true deep copy. It will copy properties from the original to the new object by reference - which should be enough for an object that will be serialized to JSON shortly thereafter anyways, but I'll leave a comment in the code to be careful about that.